tests/python/tests/test_xml_parser.py

```diff
@@ -1156,3 +1170,187 @@ def warningcb(type, msg):
+        self.assertEqual(pkg.obsoletes,
+                [('kernel', None, None, None, None, False),
+                 ('super_kernel', 'EQ', '0', '5.9.0', None, False)])
+        self.assertEqual(pkg.files,
+                [(None, '/usr/bin/', 'super_kernel'),
+                 (None, '/usr/share/man/', 'super_kernel.8.gz')])
+        self.assertEqual(pkg.changelogs,
+                [('Tomas Mlcoch <tmlcoch@redhat.com> - 6.0.1-1',
+                   1334664000,
+                  '- First release'),
+                 ('Tomas Mlcoch <tmlcoch@redhat.com> - 6.0.1-2',
+                   1334664001,
+                   '- Second release')])
+
+        self.assertIsNone(next(package_iterator))
+        self.assertTrue(package_iterator.is_finished())
+
+    def test_xml_parser_pkg_iterator_repo02(self):
+        warnings = []
+        def warningcb(warn_type, msg):
+            warnings.append((warn_type, msg))
+
+        package_iterator = cr.PackageIterator(
+            primary_path=REPO_02_PRIXML, filelists_path=REPO_02_FILXML, other_path=REPO_02_OTHXML,
+            warningcb=warningcb,
+        )
+
+        package_names = ['fake_bash', 'super_kernel']
+        for (pkg, name) in zip(package_iterator, package_names):
+            self.assertEqual(pkg.name, name)
+            # TODO: might be good to test that iter.is_finished() returns false while still processing packages
+
+        self.assertListEqual(warnings, [])
+        self.assertIsNone(next(package_iterator))
+        self.assertTrue(package_iterator.is_finished())
+
+    def test_xml_parser_pkg_iterator_repo02_newpkgcb_as_filter(self):
+        def newpkgcb(pkgId, name, arch):
+            if name in {"fake_bash"}:
+                return cr.Package()
+
+        package_iterator = cr.PackageIterator(
+            primary_path=REPO_02_PRIXML, filelists_path=REPO_02_FILXML, other_path=REPO_02_OTHXML,
+            newpkgcb=newpkgcb,
+        )
+
+        packages = list(package_iterator)
+
+        self.assertEqual(len(packages), 1)
+        self.assertEqual(packages[0].name, "fake_bash")
+        self.assertIsNone(next(package_iterator))
+        self.assertTrue(package_iterator.is_finished())
+
+    def test_xml_parser_pkg_iterator_warnings(self):
+        warnings = []
+        def warningcb(warn_type, msg):
+            warnings.append((warn_type, msg))
+
+        package_iterator = cr.PackageIterator(
+            primary_path=PRIMARY_MULTI_WARN_00_PATH, filelists_path=FILELISTS_MULTI_WARN_00_PATH, other_path=OTHER_MULTI_WARN_00_PATH,
+            warningcb=warningcb,
+        )
+
+        package_names = ['fake_bash', 'super_kernel']
+        for (pkg, name) in zip(package_iterator, package_names):
+            self.assertEqual(pkg.name, name)
+            # TODO: might be good to test that iter.is_finished() returns false while still processing packages
+
+        self.assertListEqual(warnings, [])
+        self.assertIsNone(next(package_iterator))
```

Comment:
The `warnings` are not empty the `assert` below is correct.
```suggestion
        self.assertRaises(StopIteration, next, package_iterator)
```